hibernate-core/src/main/java/org/hibernate/internal/StatelessSessionImpl.java

```diff
@@ -646,4 +647,20 @@ public boolean isStatelessSession() {
 		return true;
 	}
 
+	@Override
+	public void close() throws HibernateException {
+		if ( isClosed() ) {
+			LOG.trace( "Already closed" );
+			return;
+		}
+
+		if ( LOG.isTraceEnabled() ) {
+			LOG.tracef( "Closing session [%s]", getSessionIdentifier() );
+		}
+
+		getJdbcCoordinator().executeBatch();
```

Comment:
It seems `SessionImpl` has some methods which `StatelessSessionImpl` lacks that could affect what the inherited `close` method does:

```
	@Override
	public void delayedAfterCompletion() {
		TransactionCoordinator coordinator = getTransactionCoordinator();
		if ( coordinator instanceof JtaTransactionCoordinatorImpl ) {
			( (JtaTransactionCoordinatorImpl) coordinator).getSynchronizationCallbackCoordinator()
					.processAnyDelayedAfterCompletion();
		}
	}

	@Override
	protected boolean shouldCloseJdbcCoordinatorOnClose(boolean isTransactionCoordinatorShared) {
		if ( !isTransactionCoordinatorShared ) {
			return super.shouldCloseJdbcCoordinatorOnClose( isTransactionCoordinatorShared );
		}

		final ActionQueue actionQueue = getActionQueue();
		if ( actionQueue.hasBeforeTransactionActions() || actionQueue.hasAfterTransactionActions() ) {
			log.warn(
					"On close, shared Session had before/after transaction actions that have not yet been processed"
			);
		}
		return false;
	}
```

Can you please set a break point in `JdbcCoordinator.executeBatch()` to see how that method is called from `SessionImpl` and whether adding one or both methods I mentioned here to `StatelessSessionImpl` could be a "more appropriate fix".

Reply:
Tbh I don't think either of these two places are appropriate for executing any lingering batch when the session is about to be closed, at least not in any similar form to what SessionImpl does in them